Grumpkin curve implementation

nonam3e opened the pull request:
This PR adds support for the Grumpkin curve to ICICLE

ChickenLover commented on icicle/curves/grumpkin_params.cuh:
```diff
@@ -0,0 +1,95 @@
+
+  // G1 and G2 generators
+  static constexpr storage<fq_config::limbs_count> g1_gen_x = {0x00000001, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000};
+  static constexpr storage<fq_config::limbs_count> g1_gen_y = {0x823f272c, 0x833fc48d, 0xf1181294, 0x2d270d45, 0x6a45d63, 0xcf135e75, 0x2, 0x0};
```
Please pad to 4 bytes

ChickenLover commented on wrappers/rust/icicle-curves/icicle-grumpkin/Cargo.toml:
```diff
@@ -0,0 +1,28 @@
+[dependencies]
+icicle-core = { path = "../../icicle-core" }
+icicle-cuda-runtime = { path = "../../icicle-cuda-runtime" }
+ark-grumpkin-test = { git = "https://github.com/nonam3e/ark-grumpkin-test.git", optional = true}
```
Change to https://github.com/ingonyama-zk/ark-grumpkin-test

ChickenLover commented on wrappers/rust/icicle-curves/icicle-grumpkin/Cargo.toml:
```diff
@@ -0,0 +1,28 @@
+[package]
+name = "icicle-grumpkin"
+version = "1.0.0"
```
bump to 1.4.0

ChickenLover replied:
it's 1.5.0 now

ChickenLover commented on wrappers/rust/icicle-curves/icicle-grumpkin/README.md:
```diff
@@ -0,0 +1,26 @@
+### Notes
+
+- There's no g2 for Grumpkin curve
+- Should implement Grumpkin curve for tests, because ark-grumpkin doesn't work with current macroses
```
remove this line

ChickenLover commented on wrappers/rust/icicle-curves/icicle-grumpkin/src/msm/mod.rs:
```diff
@@ -0,0 +1,38 @@
+};
+use icicle_cuda_runtime::{error::CudaError, memory::HostOrDeviceSlice};
+
+// impl MSM<CurveCfg> for CurveCfg {
```
remove this section

ChickenLover reviewed (commented):
don't forget to run `./scripts/hooks/pre-push`

DmytroTym reviewed (commented):
Thanks for contributing to icicle!
Good job with the PR. In general I think it's a good idea to re-use definitions from bn254 for grumpkin as any realistic application using grumpkin will probably use bn254 as well, right? On C++ side using bn254 is easy enough but with Rust it might be trickier. Although since NTT isn't needed, there should be no issues with foreign traits on foreign types but still I can do a PR against your branch if it's unclear how to do it. Btw arkworks also re-export definitions from bn254 instead of implementing from scratch: https://github.com/arkworks-rs/algebra/blob/master/curves/grumpkin/src/fields/fq.rs and https://github.com/arkworks-rs/algebra/blob/master/curves/grumpkin/src/fields/fr.rs

nonam3e commented:
@DmytroTym Hey, thanks for your review. I've marked my pull request as a draft for now and will update it after addressing the issues raised in your review.

>Btw arkworks also re-export definitions from bn254 instead of implementing from scratch: [https://github.com/arkworks-rs/algebra/blob/master/curves/grumpkin/src/fields/fq.rs](https://github.com/arkworks-rs/algebra/blob/master/curves/grumpkin/src/fields/fq.rs) and [https://github.com/arkworks-rs/algebra/blob/master/curves/grumpkin/src/fields/fr.rs](https://github.com/arkworks-rs/algebra/blob/master/curves/grumpkin/src/fields/fr.rs)

I attempted to implement tests for the grumpkin curve using both variants, but encountered a compilation problem while trying to use them. That's why I created my own repository to test the intermediate results, if manage to use the original repo, I'll change the dependency.

DmytroTym commented:
@nonam3e didn't mean that you need to use original arkworks' version, just gave an example of how they re-export fields from bn254 crate to be used in grumpkin

DmytroTym reviewed (approved):
lgtm
